test/gurka/test_build_admin.cc

```diff
@@ -0,0 +1,183 @@
+  pt.put("mjolnir.tile_dir", "test/data/admin");
+  pt.put("mjolnir.admin", "test/data/admin.sqlite");
+  pt.put("mjolnir.timezone", "test/data/not_needed.sqlite");
+
+  // Given map.pbf, BuildAdminFromPBF() creates test/data/admin.sqlite.
+  std::vector<std::string> input_files = {workdir + "/map.pbf"};
+  BuildAdminFromPBF(pt.get_child("mjolnir"), input_files);
+
+  // reverse engineer the tile_id from the nodes that make up
+  // our mock map.pbf. Its probably overkill to check every node
+  // for its tile-id if I know they are all in the same tile-id,
+  // but its a good exercise.
+  std::unordered_set<GraphId> tile_ids;
+  for (const auto& node : admin_map.nodes) {
+    const midgard::PointLL& latlon = node.second;
+    GraphId tile_id = TileHierarchy::GetGraphId(latlon, 0);
+    tile_ids.insert(tile_id);
+  }
+  ASSERT_EQ(tile_ids.size(), 1);
+
+  GraphId tile_id(*tile_ids.begin());
+
+  // Load the admin.sqlite table we just created
+  sqlite3* db_handle = NULL;
+  std::string dbname = pt.get<std::string>("mjolnir.admin");
+  uint32_t ret = sqlite3_open_v2(dbname.c_str(), &db_handle, SQLITE_OPEN_READONLY, NULL);
+  EXPECT_EQ(ret, SQLITE_OK);
+
+  // Create a GraphReader so we can create a GraphTileBuilder
+  GraphReader graph_reader(pt.get_child("mjolnir"));
+  auto t = GraphTile::Create(graph_reader.tile_dir(), tile_id);
+  ASSERT_TRUE(t);
+  GraphTileBuilder tilebuilder(graph_reader.tile_dir(), tile_id, true);
+
+  // Call GetAdminInfo.
+  std::unordered_multimap<uint32_t, multi_polygon_type> polys;
+  std::unordered_map<uint32_t, bool> drive_on_right;
+  std::unordered_map<uint32_t, bool> allow_intersection_names;
+  AABB2<PointLL> world_box(-180.0f, -90.0f, 180.f, 90.f);
+  polys = GetAdminInfo(db_handle, drive_on_right, allow_intersection_names, world_box, tilebuilder);
+  sqlite3_close(db_handle);
+
+  // For two states part of one country, there should be three polys, etc.
+  ASSERT_EQ(polys.size(), 3);
+  ASSERT_EQ(drive_on_right.size(), 3);
+  ASSERT_EQ(allow_intersection_names.size(), 3);
+
+  // find the GH way/edge in the graph, make sure its there
+  auto tup = findEdge(graph_reader, admin_map.nodes, "GH", "H", tile_id);
+  const DirectedEdge* dir_edge = std::get<1>(tup);
+  ASSERT_NE(dir_edge, nullptr);
+
+  // See that nodes G and H are in the correct state
+  for (const auto& node : admin_map.nodes) {
+    const std::string& node_name = node.first;
+    if (node_name == "G") {
+      uint32_t co_offset = tilebuilder.AddName("Colorado");
+      const midgard::PointLL& node_latlon = node.second;
+      uint32_t state_offset = GetContainingState(tilebuilder, polys, node_latlon);
+      ASSERT_EQ(state_offset, co_offset);
+    } else if (node_name == "H") {
+      uint32_t ut_offset = tilebuilder.AddName("Utah");
+      const midgard::PointLL& node_latlon = node.second;
+      uint32_t state_offset = GetContainingState(tilebuilder, polys, node_latlon);
+      ASSERT_EQ(state_offset, ut_offset);
+    }
+  }
```

Comment:
i think we can do something like the following to test that the admins made it into the tiles where they belong with the right values:

```c++
  // get the edge in the direction of G to H and the opp edge in the direction of H to G
  const DirectedEdge* edge, *opp_edge;
  std::tie(std::ignore, edge, std::ignore, opp_edge) = gurka::findEdge(*reader, map.nodes, "GH", "H");
  // grab the end node of the edge (that should be H)
  auto tile = reader->GetGraphTile(edge->endnode());
  const auto* H = tile->node(edge->endnode());
  // get also the end node of the opp edge that should be G
  auto opp_tile = reader->GetGraphTile(opp_edge->endnode());
  const auto* G = opp_tile->node(opp_edge->endnode());
  // figure out what the admin info for H and G
  auto H_admin = tile->admininfo(H->admin_index());
  auto G_admin = tile->admininfo(H->admin_index());
  //TODO: some asserts that it matches what was in the pbf
  ```

Reply:
Thanks, suggestions taken.